### ISSUE TITLE
Segmentation Fault s3_get

### ISSUE THREAD

**Issue by mahiki** [JuliaCloud/AWSS3.jl#213]
I'm able to do AWSS3.read of s3 path to object, but in the same project I'm able to repeatedly hit a seg fault when using `AWS.s3_get`. Here is the error:

```jl
using AWSS3, AWS
aws = global_aws_config(; region = "us-east-1")
# AWSConfig((AKIA??????????H6, I5j..., 146138512-12-31T23:59:59), "us-east-1", "json")

AWSS3.s3_get(aws, "test-data", "test/test-37/s3-walkpath", "df-data-1.txt")
# zsh: segmentation fault  julia --project=. --startup-file=no

# using the correct number of arguments had no problem
AWSS3.s3_get(aws, "test-data", "test/test-37/s3-walkpath/df-data-1.txt")

# this is correct object contents:
#=
"column_1\tamount\titem\tday\textra\n0\t19.0\tB001\t2021-01-01\tHello newfile\n1\t11.0\tB001\t2021-01-01\tHello newfile\n0\t35.5\tB020\t2112-12-12\tHello newfile\n1\t32.5\tB020\t2020-10-20\tHello newfile\n0\t5.99\tBX00\t2021-05-04\tHello newfile\n1\t5.99\tBX00\t1984-07-04\tHello newfile\n"
=#
```

I mistakenly passed the "bucket", "path", "filename" instead of "bucket", "path/filename" arguments, because I was thinking of the AWSS3.s3_get_file() call which takes "filename" as a separate argument.

Here is the project configuration:

```jl
pkg> st
  [fbe9abb3] AWS v1.61.0
  [1c724243] AWSS3 v0.9.2

```

-------
I also find I was confused by the documentation of s3_get_file, resulted in me chasing around the error stacktrace. The documentation is clear, once I realize my mistake, but `target_filename`  or `local_filename` (as in the awscli documentation) might be better than `filename`?

```jl
help?> AWSS3.s3_get_file
  s3_get_file([::AbstractAWSConfig], bucket, path, filename; [version=], kwargs...)
```

It wasn't immediately obvious "filename" is the *target* filename on the local system, and "path" is key to the object in s3.

The AWS CLI doc looks like:
```
aws s3 cp help
SYNOPSIS
            cp
          <LocalPath> <S3Uri> or <S3Uri> <LocalPath> or <S3Uri> <S3Uri>
```

**Comment by mahiki** [comment]
Yes mac.  I'm just a data consumer, not really sure the difference of seg fault vs stack overflow.  At any rate it seems to me the kwargs handling is taking the extra argument as something unexpected.

Is it idiomatic to do some error catching or argument checking in julia functions?

**Comment by ericphanson** [comment]
So a stack overflow generally means you called a function which called a function which called a function which .... so many times the "stack", which is an internal data structure inside the programming language itself, ran out of room to keep track of where we are. Usually it means you have some unintentional recursion, which is exactly what happens here: we end up calling
```julia
 s3_get(a...; b...) = s3_get(global_aws_config(), a...; b...) 
```
over and over and over when there's an extra argument, because that method accepts any number of arguments.

The fix is to not write methods like that :). The point of that function is to add in the config if the user doesn't specify a config. So we can just write it as
```julia
 s3_get(bucket, key; b...) = s3_get(global_aws_config(), bucket, key; b...) 
```
Now that function only accepts 2 positional arguments, `bucket` and `key`. If you pass three, it will try to apply the method of that function which accepts three positional arguments:
https://github.com/JuliaCloud/AWSS3.jl/blob/5e73e8e2ca6e5fa49023cb1ccf39586b9e7cb6ed/src/AWSS3.jl#L98-L109
if the first one is not a config object, it will throw a simple error.

That's why I've labelled this as a "good first issue": since it has a simple fix, which is just replacing the line

https://github.com/JuliaCloud/AWSS3.jl/blob/5e73e8e2ca6e5fa49023cb1ccf39586b9e7cb6ed/src/AWSS3.jl#L136

by
```julia
 s3_get(bucket, key; b...) = s3_get(global_aws_config(), bucket, key; b...) 
```

(Note: in the discussion above I didn't talk about keyword arguments at all-- the reason is that they don't participate in multiple dispatch in Julia, which means we don't choose which method to call based on the keyword arguments; instead, Julia only uses the positional arguments).

**Comment by mahiki** [comment]
After a few readings, this is actually kind of humorous. Sorry.

I am sure the group will solve the issue properly. I feel there is some CS concept at work here that I got pulled into unknowingly. It seems weird this is an outcome?  I just want to learn enough not to create this situation for packages I create in the future.

Thanks!
